File: lib/features/operations/domain/observability.dart
enum LogLevel { debug, info, warning, error, critical }

enum HealthStatus { healthy, degraded, unavailable }

enum IncidentStatus { open, acknowledged, investigating, resolved }

class ApplicationLog {
  const ApplicationLog({
    required this.id,
    required this.level,
    required this.service,
    required this.message,
    required this.timestamp,
    required this.correlationId,
    required this.context,
  });
  final String id;
  final LogLevel level;
  final String service;
  final String message;
  final DateTime timestamp;
  final String correlationId;
  final Map<String, Object?> context;
}

class MetricPoint {
  const MetricPoint({
    required this.name,
    required this.value,
    required this.unit,
    required this.timestamp,
    required this.labels,
  });
  final String name;
  final double value;
  final String unit;
  final DateTime timestamp;
  final Map<String, String> labels;
}

class TraceSpan {
  const TraceSpan({
    required this.traceId,
    required this.spanId,
    required this.operation,
    required this.service,
    required this.startedAt,
    required this.duration,
    required this.successful,
    this.parentSpanId,
  });
  final String traceId;
  final String spanId;
  final String? parentSpanId;
  final String operation;
  final String service;
  final DateTime startedAt;
  final Duration duration;
  final bool successful;
}

class ServiceHealth {
  const ServiceHealth({
    required this.service,
    required this.status,
    required this.checkedAt,
    required this.latencyMilliseconds,
    required this.details,
  });
  final String service;
  final HealthStatus status;
  final DateTime checkedAt;
  final int latencyMilliseconds;
  final Map<String, Object?> details;
}

class AlertRule {
  const AlertRule({
    required this.id,
    required this.metricName,
    required this.threshold,
    required this.comparison,
    required this.severity,
    required this.escalationTarget,
    required this.enabled,
  });
  final String id;
  final String metricName;
  final double threshold;
  final String comparison;
  final LogLevel severity;
  final String escalationTarget;
  final bool enabled;
}

class OperationalIncident {
  const OperationalIncident({
    required this.id,
    required this.title,
    required this.severity,
    required this.status,
    required this.createdAt,
    required this.escalationTarget,
    required this.notes,
  });
  final String id;
  final String title;
  final LogLevel severity;
  final IncidentStatus status;
  final DateTime createdAt;
  final String escalationTarget;
  final List<String> notes;
}

class PerformanceReport {
  const PerformanceReport({
    required this.requestVolume,
    required this.errorRate,
    required this.averageResponseMilliseconds,
    required this.metrics,
    required this.generatedAt,
  });
  final int requestVolume;
  final double errorRate;
  final double averageResponseMilliseconds;
  final Map<String, double> metrics;
  final DateTime generatedAt;
}

File: scholarsphere_backend/app/models/observability.py
import enum
import uuid
from datetime import datetime
from typing import Any

from sqlalchemy import Boolean, DateTime, Enum, Float, String, Text
from sqlalchemy.orm import Mapped, mapped_column
from sqlalchemy.types import Uuid

from app.db.base import Base
from app.models.external_opportunity import JSONType


class LogLevel(str, enum.Enum):
    debug = "debug"
    info = "info"
    warning = "warning"
    error = "error"
    critical = "critical"


class HealthStatus(str, enum.Enum):
    healthy = "healthy"
    degraded = "degraded"
    unavailable = "unavailable"


class IncidentStatus(str, enum.Enum):
    open = "open"
    acknowledged = "acknowledged"
    investigating = "investigating"
    resolved = "resolved"


class ApplicationLog(Base):
    __tablename__ = "application_logs"

    id: Mapped[uuid.UUID] = mapped_column(Uuid, primary_key=True, default=uuid.uuid4)
    level: Mapped[LogLevel] = mapped_column(Enum(LogLevel, native_enum=False), nullable=False)
    service: Mapped[str] = mapped_column(String(255), nullable=False)
    message: Mapped[str] = mapped_column(Text, nullable=False)
    timestamp: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    correlation_id: Mapped[str] = mapped_column(String(255), nullable=False)
    context: Mapped[dict[str, Any]] = mapped_column(JSONType, default=dict, nullable=False)


class MetricPoint(Base):
    __tablename__ = "metric_points"

    id: Mapped[uuid.UUID] = mapped_column(Uuid, primary_key=True, default=uuid.uuid4)
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    value: Mapped[float] = mapped_column(Float, nullable=False)
    unit: Mapped[str] = mapped_column(String(64), nullable=False)
    timestamp: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    labels: Mapped[dict[str, str]] = mapped_column(JSONType, default=dict, nullable=False)


class TraceSpan(Base):
    __tablename__ = "trace_spans"

    trace_id: Mapped[str] = mapped_column(String(255), primary_key=True)
    span_id: Mapped[str] = mapped_column(String(255), primary_key=True)
    parent_span_id: Mapped[str | None] = mapped_column(String(255), nullable=True)
    operation: Mapped[str] = mapped_column(String(255), nullable=False)
    service: Mapped[str] = mapped_column(String(255), nullable=False)
    started_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    duration_seconds: Mapped[float] = mapped_column(Float, nullable=False)
    successful: Mapped[bool] = mapped_column(Boolean, nullable=False)


class AlertRule(Base):
    __tablename__ = "alert_rules"

    id: Mapped[str] = mapped_column(String(255), primary_key=True)
    metric_name: Mapped[str] = mapped_column(String(255), nullable=False)
    threshold: Mapped[float] = mapped_column(Float, nullable=False)
    comparison: Mapped[str] = mapped_column(String(32), nullable=False)
    severity: Mapped[LogLevel] = mapped_column(Enum(LogLevel, native_enum=False), nullable=False)
    escalation_target: Mapped[str] = mapped_column(String(255), nullable=False)
    enabled: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)


class OperationalIncident(Base):
    __tablename__ = "operational_incidents"

    id: Mapped[str] = mapped_column(String(255), primary_key=True)
    title: Mapped[str] = mapped_column(String(1000), nullable=False)
    severity: Mapped[LogLevel] = mapped_column(Enum(LogLevel, native_enum=False), nullable=False)
    status: Mapped[IncidentStatus] = mapped_column(
        Enum(IncidentStatus, native_enum=False), nullable=False
    )
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    escalation_target: Mapped[str] = mapped_column(String(255), nullable=False)
    notes: Mapped[list[str]] = mapped_column(JSONType, default=list, nullable=False)
